packages/office-ui-fabric-react/src/components/SpinButton/SpinButton.tsx

```diff
@@ -61,60 +76,55 @@ export class SpinButton extends BaseComponent<ISpinButtonProps, ISpinButtonState
     decrementButtonIcon: { iconName: 'ChevronDownSmall' }
   };
 
-  private _input = React.createRef<HTMLInputElement>();
-  private _inputId: string;
-  private _labelId: string;
-  private _lastValidValue: string;
-  private _spinningByMouse: boolean;
-  private _valueToValidate: string | undefined; // To avoid duplicate validations/submissions
+  private readonly _input = React.createRef<HTMLInputElement>();
+  private readonly _labelId: string = getId('label');
+  private readonly _inputId: string = getId('input');
+
+  // @todo The `_precision` should be a memorized getter on `props`.
   private _precision: number;
 
-  private _currentStepFunctionHandle: number;
-  private _initialStepDelay = 400;
-  private _stepDelay = 75;
+  private _currentStepFunctionHandle: number = -1;
+  private readonly _initialStepDelay = 400;
+  private readonly _stepDelay = 75;
 
-  constructor(props: ISpinButtonProps) {
+  public constructor(props: ISpinButtonProps) {
     super(props);
 
     this._warnMutuallyExclusive({
       value: 'defaultValue'
     });
 
     const value = props.value || props.defaultValue || String(props.min) || '0';
-    this._lastValidValue = value;
 
-    // Ensure that the autocalculated precision is not negative.
+    // Ensure that the auto-calculated precision is not negative.
     this._precision = this._calculatePrecision(this.props as ISpinButtonInternalProps);
 
     this.state = {
       isFocused: false,
       value: value,
-      keyboardSpinDirection: KeyboardSpinDirection.notSpinning
+      lastValidValue: value, // @todo Can we use this._validateValue(value)? Will it break the validate invoke assumption?
+      spinDirection: SpinDirection.notSpinning,
+      keyboardSpinDirection: SpinDirection.notSpinning
     };
-
-    this._currentStepFunctionHandle = -1;
-    this._labelId = getId('Label');
-    this._inputId = getId('input');
-    this._spinningByMouse = false;
-    this._valueToValidate = undefined;
   }
 
-  /**
-   * Invoked when a component is receiving new props. This method is not called for the initial render.
-   */
-  public componentWillReceiveProps(newProps: ISpinButtonProps): void {
-    this._lastValidValue = this.state.value;
-    let value: string = newProps.value ? newProps.value : String(newProps.min);
-    if (newProps.defaultValue) {
-      value = String(Math.max(newProps.min as number, Math.min(newProps.max as number, Number(newProps.defaultValue))));
+  public componentWillReceiveProps(nextProps: ISpinButtonProps): void {
+    if (this.props.value !== undefined && nextProps.value !== undefined) {
+      // Controlled mode, respect the next value if it is changed.
+      const currValue: string = this.props.value || '0';
+      const nextValue: string = nextProps.value || '0';
+      if (currValue !== nextValue) {
+        this.setState({
+          value: nextValue
```

Comment:
lastValidValue should have also been updated here. Without updating it, I believe pressing escape would revert to the previous value.